### PR TITLE
Weather future

### PR DESCRIPTION
**Pull request by KonstantinZhukovskij**

### CODE REVIEW

File: client/src/components/MainPage/index.js [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+                        isLoading: false,
+                        isDisplayMap: true,
+                        isDisplayFutureWeather: false
+                    });
+                    let startPlace = res.data.response.GeoObjectCollection.featureMember;
+                    startPlace.map((item) => {
+                        return startPlace = item.GeoObject.Point.pos
+                    });
+                    this.setState({
+                        longitude: startPlace.split(' ')[0],
```
[the remark this answers is not in the document]
Comment (nikolyachikm) [reply client/src/components/MainPage/index.js]: look, you should call setState only twice for this method - 
1st - set isLoading: true, 
2nd - isLoading:false and some data or error

in your case you call it 4 times in good scenario

File: client/src/components/CityDetails/style.css [hunk 2]
```diff
@@ -0,0 +1,3 @@
+.cityDetails {
+    margin-bottom: 1%;
+}
```
[the remark this answers is not in the document]
Comment (KonstantinZhukovskij) [reply client/src/components/CityDetails/style.css]: +

File: client/src/components/CloudTooltip/index.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+export default class CloudTooltip extends Component {
+    render() {
+        return (
+            <div className="cloud">
+                <div className="cloudShadow">
+                    <h5>Temperature: {this.props.temperatureToday}°С</h5>
+                    <h5>Humidity: {this.props.humidityToday} %</h5>
+                    <h5>Wind Speed: {this.props.windSpeedToday} m/s</h5>
+                    <button className="btn detailsButton" onClick={this.props.onClickRenderWeatherDetails}>Next Days
```
[the remark this answers is not in the document]
Comment (KonstantinZhukovskij) [reply client/src/components/CloudTooltip/index.js]: Done

File: client/src/components/MainPage/index.js [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+                    </div>
+                </div>
+                {this.state.isLoading && this.showSpinner()}
+                {this.state.isWeatherInfoReceived && this.state.isDisplayMap && this.renderWeatherIfThereData()}
+                {this.state.isDisplayFutureWeather && this.RenderWeatherDetailsForFuture()}
+            </div>
+        )
+    }
+
+    static propTypes = {
```
[the remark this answers is not in the document]
Comment (KonstantinZhukovskij) [reply client/src/components/MainPage/index.js]: propTypes Done. Component.propTypes doesn't work

File: client/src/components/MainPage/index.js [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+                isDisplayFutureWeather: false
+            });
+        } else {
+            this.setState({
+                isDisplayFutureWeather: true
+            });
+        }
+    };
+
+    RenderWeatherDetailsForFuture = () => {
```
[the remark this answers is not in the document]
Comment (KonstantinZhukovskij) [reply client/src/components/MainPage/index.js]: Done

File: client/src/components/MainPage/index.js [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+    };
+
+    showSpinner = () => {
+        return (
+            <Spinner/>
+        )
+    };
+
+    onClickRenderWeatherDetails = () => {
+        if (this.state.isDisplayFutureWeather) {
```
[the remark this answers is not in the document]
Comment (KonstantinZhukovskij) [reply client/src/components/MainPage/index.js]: Done

File: client/src/components/MainPage/index.js [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+    };
+
+    render() {
+        return (
+            <div>
+                <div className='row'>
+                    <div className='col-12'>
+                        <SearchForm city={this.state.city}
+                                    onCityChange={this.onCityChange}
+                                    getWeather={this.onClickGetWeather}/>
```
[the remark this answers is not in the document]
Comment (KonstantinZhukovskij) [reply client/src/components/MainPage/index.js]: Done

File: client/src/components/Popup/index.js [hunk 8]
```diff
@@ -1,18 +1,18 @@
...
                 <div className='modalContent'>
-                    <div className="row">
-                        <div className="col-12">
+                    <div className='row'>
+                        <div className='col-12'>
                             <span className='closeButton'>&times;</span>
                         </div>
                     </div>
-                    <ModelContent/>
+                    <MainPage/>
```
[the remark this answers is not in the document]
Comment (KonstantinZhukovskij) [reply client/src/components/Popup/index.js]: Done

File: client/src/components/Spinner/index.js [hunk 9]
```diff
@@ -4,13 +4,13 @@ import './style.css';
 const Spinner = () => {
     return (
         <div className='spinner'>
-            <div className="lds-css ng-scope">
-                <div className="lds-eclipse">
-                    <div></div>
+            <div className='lds-css ng-scope'>
+                <div className='lds-eclipse'>
+                    <div>
```
[the remark this answers is not in the document]
Comment (KonstantinZhukovskij) [reply client/src/components/Spinner/index.js]: IDE doesn't underline my version ))

File: client/src/components/MainPage/index.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (KonstantinZhukovskij) [reply client/src/components/MainPage/index.js]: Done

File: client/src/components/MainPage/index.js [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+            isLoading: false
+        });
+        // getAutoComplete(this.state.city)Spinner
+        //     .then((res) => {
+        //         console.log(res)
+        //     })
+    };
+
+    onClickGetWeather = () => {
+        if (this.state.city !== '') {
```
[the remark this answers is not in the document]
Comment (KonstantinZhukovskij) [reply client/src/components/MainPage/index.js]: Done

File: client/src/components/MainPage/index.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (KonstantinZhukovskij) [reply client/src/components/MainPage/index.js]: This is ES7 version
static propTypes = {
    ...
  }

File: client/src/components/YandexMap/index.js [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+    closePopup = (event) => {
+        this.setState({isDisplayPopup: false});
+    };
+
+    closePopupWithEsc = (event) => {
+        (event.keyCode === 27) && this.setState({isDisplayPopup: false});
+    };
+
+    render() {
+        const placeMark = {
```
Comment (nikolyachikm) [inline client/src/components/YandexMap/index.js]: move to separate function

File: client/src/App.css [hunk 12]
```diff
@@ -0,0 +1,236 @@
+body {
```
[the remark this answers is not in the document]
Comment (KonstantinZhukovskij) [reply client/src/App.css]: No, it's inside

[IMG](https://imgur.com/a/IWZbeIz)

File: client/src/components/MainPage/index.js [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+            return this.state.city
+        }
+    };
+
+    getWeather = () => {
+        if (!this.validateSearchValue(this.state.city)) {
+            toastr.warning('It seems you have entered an invalid name or left input blank');
+            return
+        }
+        this.setState({
```
[the remark this answers is not in the document]
Comment (KonstantinZhukovskij) [reply client/src/components/MainPage/index.js]: Done

File: client/src/components/MainPage/index.js [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+            isLoading: true
+        });
+        getPlace(this.state.city)
+            .then((res) => {
+                const {featureMember} = res.data.response.GeoObjectCollection;
+                const cityName = featureMember[0].GeoObject.name;
+                const locationCoordinates = featureMember[0].GeoObject.Point.pos;
+                localStorage.setItem(cityName, locationCoordinates);
```
[the remark this answers is not in the document]
Comment (KonstantinZhukovskij) [reply client/src/components/MainPage/index.js]: Done

File: client/src/components/MainPage/index.js [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+            isLoading: false
+        })
+    };
+
+    validateSearchValue = (city) => {
+        const cityName = city.trim().length;
+        const regExp = /^[A-Za-zА-Яа-яЁё\s-]+$/;
+        if (cityName > 2 && regExp.test(city)) {
+            return this.state.city
+        }
```
[the remark this answers is not in the document]
Comment (KonstantinZhukovskij) [reply client/src/components/MainPage/index.js]: Done. Added `else {
            return false
        }`

File: client/src/components/MainPage/index.js [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+    getWeather = () => {
+        if (!this.validateSearchValue(this.state.city)) {
+            toastr.warning('It seems you have entered an invalid name or left input blank');
+            return
+        }
+        this.setState({
+            isLoading: true
+        });
+        getPlace(this.state.city)
+            .then((res) => {
```
[the remark this answers is not in the document]
Comment (KonstantinZhukovskij) [reply client/src/components/MainPage/index.js]: Done

File: client/src/components/YandexMap/index.js (hunk 11 above)
Comment (KonstantinZhukovskij) [reply client/src/components/YandexMap/index.js]: Done. Transferred to separate functions getPlaceMark and getLastCities

File: client/src/components/MainPage/index.js [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+                const {featureMember} = res.data.response.GeoObjectCollection;
```
[the remark this answers is not in the document]
Comment (KonstantinZhukovskij) [reply client/src/components/MainPage/index.js]: Done

File: client/src/components/CloudBackground/index.js [hunk 18]
```diff
@@ -0,0 +1,35 @@
+import React, {Component} from 'react';
+import './style.css';
+
+export default class CloudBackground extends Component {
+    render() {
+        return (
+            <div id="background-wrap">
+                <div className="x1">
```
[the remark this answers is not in the document]
Comment (KonstantinZhukovskij) [reply client/src/components/CloudBackground/index.js]: Done

File: client/src/App.scss [hunk 19]
```diff
@@ -0,0 +1,287 @@
+// CloudBackground
+
```
[the remark this answers is not in the document]
Comment (KonstantinZhukovskij) [reply client/src/App.scss]: Done

File: client/src/components/MainPage/index.js [hunk 20]
```diff
@@ -0,0 +1,161 @@
...
+            })
+            .catch((error) => {
+                console.log('getWeather error:', error);
+            })
+    };
+
+    validateSearchValue = (city) => {
+        const cityName = city.trim().length;
+        const regExp = /^[A-Za-zА-Яа-яЁё\s-]+$/;
+        if (cityName > 2 && regExp.test(city)) {
```
Comment (nikolyachikm) [inline client/src/components/MainPage/index.js]: return cityName > 2 && regExp.test(city);

File: client/src/components/MainPage/index.js [hunk 21]
```diff
@@ -0,0 +1,161 @@
...
+    getPlace = () => {
+        return getPlaceData(this.state.city)
+            .then((response) => {
+                const { featureMember } = response.data.GeoObjectCollection;
+                const lastCityIndex = localStorage.length;
+                const cityName = lastCityIndex + ' ' + featureMember[0].GeoObject.name;
+                const locationCoordinates = featureMember[0].GeoObject.Point.pos;
+                localStorage.setItem(cityName, locationCoordinates);
+
+                this.setState({
```
Comment (nikolyachikm) [inline client/src/components/MainPage/index.js]: get... method should not set anything

File: client/src/components/MainPage/index.js [hunk 22]
```diff
@@ -0,0 +1,161 @@
...
+        if (!this.validateSearchValue(this.state.city)) {
+            toastr.warning('It seems you have entered an invalid name or left input blank');
+            return
+        }
+
+        this.setState({ isLoading: true });
+
+        this.getPlace()
+            .then(() => {
+                this.getWeather()
```
Comment (nikolyachikm) [inline client/src/components/MainPage/index.js]: move setState from getPlace and getWeather here

File: client/src/components/MainPage/index.js [hunk 23]
```diff
@@ -0,0 +1,161 @@
...
+            })
+    };
+
+    showMapIfUserWasEarlier = () => {
+        if (localStorage.length >= 1) {
+            this.setState({
+                isWeatherInfoReceived: true,
+                isDisplayMap: true,
+                city: localStorage.key(localStorage.length - 1).split(' ')[1],
+                latitude: localStorage.getItem(localStorage.key(localStorage.length - 1)).split(' ')[1],
```
Comment (nikolyachikm) [inline client/src/components/MainPage/index.js]: what if my local storage contains much data not realted to your site? 
so
1) move getting city and latitude to top of function
2) add your uniq key to localStorage

File: server/controllers/index.js [hunk 24]
```diff
@@ -0,0 +1,28 @@
...
+    return axios.get(`https://api.darksky.net/forecast/19a7a53f5d56338fe47843b7d03a3fdd/${latitude}, ${longitude}?units=auto`)
+        .then((weather) => {
+            return response.json(weather.data)
+        })
+        .catch((error) => {
+            console.log("error: ", error)
+        })
+};
+
+
```
Comment (nikolyachikm) [inline server/controllers/index.js]: too many empty lines
also check other files to contain empty line at the end of file (enable rule in linter)

File: server/routers/index.js [hunk 25]
```diff
@@ -0,0 +1,9 @@
+const express = require('express');
+const weatherController = require('../controllers');
+
+const router = express.Router();
+
+router.get('/get-place-data', weatherController.getPlaceData);
```
Comment (nikolyachikm) [inline server/routers/index.js]: no need to move in separrate controller now

File: server/app.js [hunk 26]
```diff
@@ -0,0 +1,19 @@
...
+const cors = require('cors');
+const bodyParser = require('body-parser');
+const routers = require('./routers');
+
+const app = express();
+
+app.use(cors());
+app.set('port', process.env.PORT || 3001);
+
+app.use(bodyParser.json());
```
Comment (nikolyachikm) [inline server/app.js]: why you need bodyParser ?

File: client/src/components/MainPage/index.js (hunk 20 above)
Comment (KonstantinZhukovskij) [reply client/src/components/MainPage/index.js]: Done

File: client/src/components/MainPage/index.js (hunk 21 above)
Comment (KonstantinZhukovskij) [reply client/src/components/MainPage/index.js]: Done

File: client/src/components/MainPage/index.js (hunk 22 above)
Comment (KonstantinZhukovskij) [reply client/src/components/MainPage/index.js]: Done

File: server/controllers/index.js (hunk 24 above)
Comment (KonstantinZhukovskij) [reply server/controllers/index.js]: Done

File: client/src/components/MainPage/index.js (hunk 23 above)
Comment (KonstantinZhukovskij) [reply client/src/components/MainPage/index.js]: Done

File: server/app.js (hunk 26 above)
Comment (KonstantinZhukovskij) [reply server/app.js]: Done

File: server/routers/index.js (hunk 25 above)
Comment (KonstantinZhukovskij) [reply server/routers/index.js]: Done

File: client/src/components/MainPage/index.js [hunk 27]
```diff
@@ -0,0 +1,144 @@
...
+
+        getPlaceData(this.state.city)
+            .then((response) => {
+                const { featureMember } = response.data.GeoObjectCollection;
+                const lastCityIndex = localStorage.length;
+                const cityName = lastCityIndex + ' ' + featureMember[0].GeoObject.name;
+                let locationCoordinates = featureMember[0].GeoObject.Point.pos;
+                localStorage.setItem(cityName, locationCoordinates);
+
+                getWeatherData(locationCoordinates.split(' ')[1], locationCoordinates.split(' ')[0])
```
Comment (nikolyachikm) [inline client/src/components/MainPage/index.js]: const latitude = locationCoordinates.split(' ')[1] ;
const longitude = ...
getWeatherData(latitude,longitude....
 this.setState({
....
latitude,
longitude,
...

File: client/src/components/MainPage/index.js (hunk 23 above)
Comment (nikolyachikm) [reply client/src/components/MainPage/index.js]: i can't see that it's done

File: client/src/components/MainPage/index.js [hunk 28]
```diff
@@ -0,0 +1,144 @@
...
+                console.log('getPlace error:', error);
+                toastr.warning('City not found');
+                this.setState({
+                    isLoading: false
+                })
+            })
+    };
+
+    showMapIfUserWasEarlier = () => {
+        if (localStorage.length >= 1) {
```
Comment (nikolyachikm) [inline client/src/components/MainPage/index.js]: i think it's better :
this.setState({ loading: true }):

const city = ....
const latitude = ...
const longitude= ...

getWeatherData(latitude , longitude).then((response) => {
setState({ loading: false, ...})
})

File: client/src/components/MainPage/index.js (hunk 27 above)
Comment (KonstantinZhukovskij) [reply client/src/components/MainPage/index.js]: Done

File: client/src/components/MainPage/index.js (hunk 28 above)
Comment (KonstantinZhukovskij) [reply client/src/components/MainPage/index.js]: Done

File: client/src/components/MainPage/index.js (hunk 23 above)
Comment (KonstantinZhukovskij) [reply client/src/components/MainPage/index.js]: Done

**Pull request merged**